File: android/app/src/main/java/com/nmic/autocorrect/Grammar.kt
package com.nmic.autocorrect

/**
 * Instant, local punctuation and spacing fixes applied as the user types, ported from the web app's lib/grammar.ts.
 * `fix(text, p, lang)` looks at the character just typed at index p and returns a small replacement near it, or null.
 */
object Grammar {
    data class Fix(val start: Int, val end: Int, val to: String, val note: String)

    private val ABBREV = setOf("e.g", "i.e", "etc", "vs", "mr", "mrs", "ms", "dr", "prof", "st", "no", "ca", "fx", "bl.a", "osv", "dvs", "evt", "mv", "jf", "pga", "ifm", "hhv", "inkl", "ekskl", "kl", "nr")
    private val letter = Regex("\\p{L}")
    private val lower = Regex("\\p{Ll}")
    private val letterOrDigit = Regex("[\\p{L}\\p{N}]")

    private fun isWordChar(c: Char?): Boolean = c != null && (c.isLetterOrDigit() || c == '\'' || c == '’' || c == '-' || Character.getType(c) == Character.NON_SPACING_MARK.toInt())
    private fun at(t: String, i: Int): Char? = if (i in t.indices) t[i] else null
    private fun quoteIsOpening(text: String, i: Int): Boolean {
        val paraStart = text.lastIndexOf('\n', i - 1) + 1
        var n = 0
        for (k in paraStart until i) if (text[k] == '"') n++
        return n % 2 == 0
    }
    private fun wordBefore(text: String, end: Int): String = Regex("[\\p{L}.]+$").find(text.substring(0, end.coerceIn(0, text.length)))?.value?.lowercase() ?: ""

    fun fix(text: String, p: Int, lang: String): Fix? {
        val ch = at(text, p) ?: return null
        val prev = at(text, p - 1)
        val prev2 = at(text, p - 2)

        // 0. English pronoun "i" on its own -> "I"
        if (lang == "en" && !isWordChar(ch) && prev == 'i' && !isWordChar(prev2) && prev2 != '\'' && prev2 != '’') return Fix(p - 1, p, "I", "pronoun I")
        // 1. Double space -> single
        if (ch == ' ' && prev == ' ') return Fix(p, p + 1, "", "double space")
        // 2. Space before closing punctuation
        if (ch in ",.;:!?)" && prev == ' ' && p >= 2 && prev2 != null && (letterOrDigit.matches(prev2.toString()) || prev2 in ")\"'’”")) {
            if (!(ch == '.' && prev2.isDigit())) return Fix(p - 1, p, "", "space before $ch")
        }
        // 3. Space after opening bracket or opening quote
        if (ch == ' ' && prev == '(') return Fix(p, p + 1, "", "space after (")
        if (ch == ' ' && prev == '"' && quoteIsOpening(text, p - 1)) return Fix(p, p + 1, "", "space after opening quote")
        // 4. Closing quote after a space
        if (ch == '"' && prev == ' ' && !quoteIsOpening(text, p) && p >= 2 && prev2 != null && (letterOrDigit.matches(prev2.toString()) || prev2 in ".,!?")) return Fix(p - 1, p, "", "space before closing quote")
        // 5. Missing space after , ; : ! ? . followed by a letter
        if (letter.matches(ch.toString()) && prev != null && prev in ",;:!?.") {
            val before = at(text, p - 2)
            if (before != null && letter.matches(before.toString())) {
                val wb = wordBefore(text, p - 1)
                val isUrlish = Regex("\\.(com|org|net|dk|io|co|app|ai)$", RegexOption.IGNORE_CASE).containsMatchIn("$wb.$ch") || Regex("^(www|http)", RegexOption.IGNORE_CASE).containsMatchIn(wb)
                if (prev == '.' && (ABBREV.contains(wb.trimEnd('.')) || wb.length <= 2 || isUrlish)) return null
                if (prev != '.' || wb.length > 2) return Fix(p, p, " ", "space after $prev")
            }
        }
        // 6. Missing space after closing bracket before a letter
        if (letter.matches(ch.toString()) && prev == ')') return Fix(p, p, " ", "space after )")
        // 7. Missing space before an opening bracket
        if (ch == '(' && prev != null && letterOrDigit.matches(prev.toString())) return Fix(p, p, " ", "space before (")
        // 8. Capital letter after sentence end, or at paragraph start
        if (lower.matches(ch.toString())) {
            val before = text.substring(0, p)
            val m = Regex("([.!?])\\s+$").find(before)
            if (m != null) {
                val wb = wordBefore(before, before.length - m.value.length)
                if (m.groupValues[1] != "." || !(ABBREV.contains(wb) || wb.length <= 1)) return Fix(p, p + 1, ch.uppercase(), "capital after sentence end")
            } else if (before.isBlank() || Regex("\\n\\s*$").containsMatchIn(before)) return Fix(p, p + 1, ch.uppercase(), "capital at paragraph start")
        }
        // 9. Doubled comma or semicolon
        if ((ch == ',' && prev == ',') || (ch == ';' && prev == ';')) return Fix(p, p + 1, "", "double $ch")
        return null
    }
}

File: lib/grammar.ts
// Instant, local punctuation and spacing fixes applied as the user types. No API calls.
// Each fix is a small replacement near the caret; the caller applies it and logs it as a "grammar" change.

export type GrammarFix = { start: number; end: number; to: string; note: string };

const ABBREV = new Set(["e.g", "i.e", "etc", "vs", "mr", "mrs", "ms", "dr", "prof", "st", "no", "ca", "fx", "bl.a", "osv", "dvs", "evt", "mv", "jf", "pga", "ifm", "hhv", "inkl", "ekskl", "ca", "kl", "nr"]);

/** Given the text after a single typed character at index `p` (so text[p] is the new char), return a fix or null. */
export function grammarFix(text: string, p: number, lang: "en" | "da" = "en"): GrammarFix | null {
  const ch = text[p];
  const prev = text[p - 1];
  const prev2 = text[p - 2];

  // 0. English pronoun "i" on its own -> "I" (checked when the following boundary is typed).
  if (lang === "en" && !isWordCharLocal(ch) && prev === "i" && !isWordCharLocal(prev2) && !(prev2 === "'" || prev2 === "’")) {
    return { start: p - 1, end: p, to: "I", note: "pronoun I" };
  }

  // 1. Double space -> single space.
  if (ch === " " && prev === " ") return { start: p, end: p + 1, to: "", note: "double space" };

  // 2. Space before closing punctuation: "word ," -> "word,"
  if (/[,.;:!?)]/.test(ch) && prev === " " && p >= 2 && /[\p{L}\p{N})"'’”]/u.test(prev2 ?? "")) {
    if (!(ch === "." && /\p{N}/u.test(prev2 ?? ""))) return { start: p - 1, end: p, to: "", note: `space before ${ch}` };
  }

  // 3. Space after opening bracket or opening quote: "( word" -> "(word", "\" word" -> "\"word"
  if (ch === " " && prev === "(") return { start: p, end: p + 1, to: "", note: "space after (" };
  if (ch === " " && prev === '"' && quoteIsOpening(text, p - 1)) return { start: p, end: p + 1, to: "", note: "space after opening quote" };

  // 4. Closing quote after a space: "word \"" -> "word\""
  if (ch === '"' && prev === " " && !quoteIsOpening(text, p) && p >= 2 && /[\p{L}\p{N}.,!?]/u.test(prev2 ?? "")) {
    return { start: p - 1, end: p, to: "", note: "space before closing quote" };
  }

  // 5. Missing space after , ; : ! ? . followed by a letter: "word,word" -> "word, word"
  if (/\p{L}/u.test(ch) && /[,;:!?.]/.test(prev ?? "")) {
    const before = text[p - 2] ?? "";
    if (/\p{L}/u.test(before)) {
      // not decimals/abbreviations/urls: "3.5", "e.g.", "example.com"
      const wordBefore = text.slice(0, p - 1).match(/[\p{L}.]+$/u)?.[0].toLowerCase() ?? "";
      const isUrlish = /\.(com|org|net|dk|io|co|app|ai)$/i.test(wordBefore + "." + ch) || /^(www|http)/i.test(wordBefore);
      if (prev === "." && (ABBREV.has(wordBefore.replace(/\.$/, "")) || wordBefore.length <= 2 || isUrlish)) return null;
      if (prev !== "." || wordBefore.length > 2) return { start: p, end: p, to: " ", note: `space after ${prev}` };
    }
  }

  // 6. Missing space after closing bracket before a letter: ")word" -> ") word"
  if (/\p{L}/u.test(ch) && prev === ")") return { start: p, end: p, to: " ", note: "space after )" };

  // 7. Missing space before an opening bracket: "word(" -> "word ("
  if (ch === "(" && /[\p{L}\p{N}]/u.test(prev ?? "")) return { start: p, end: p, to: " ", note: "space before (" };

  // 8. Capital letter after sentence end: "end. word" -> "end. Word" (also first letter of a paragraph)
  if (/\p{Ll}/u.test(ch)) {
    const before = text.slice(0, p);
    const m = /([.!?])\s+$/.exec(before);
    if (m) {
      const wordBefore = before.slice(0, before.length - m[0].length).match(/[\p{L}.]+$/u)?.[0].toLowerCase() ?? "";
      if (m[1] !== "." || !(ABBREV.has(wordBefore) || wordBefore.length <= 1)) return { start: p, end: p + 1, to: ch.toUpperCase(), note: "capital after sentence end" };
    } else if (/^\s*$/.test(before) || /\n\s*$/.test(before)) {
      return { start: p, end: p + 1, to: ch.toUpperCase(), note: "capital at paragraph start" };
    }
  }

  // 9. Doubled comma or semicolon.
  if ((ch === "," && prev === ",") || (ch === ";" && prev === ";")) return { start: p, end: p + 1, to: "", note: `double ${ch}` };

  return null;
}

/** Whether the straight quote at index i opens a quotation (even number of quotes before it in the paragraph). */
function quoteIsOpening(text: string, i: number): boolean {
  const paraStart = text.lastIndexOf("\n", i - 1) + 1;
  let n = 0;
  for (let k = paraStart; k < i; k++) if (text[k] === '"') n++;
  return n % 2 === 0;
}

function isWordCharLocal(c: string | undefined): boolean {
  return !!c && /[\p{L}\p{M}\p{N}'’\-]/u.test(c);
}
